### PR TITLE
2 Listeners & sasl scram & improvements on sasl_plain

### PR DESCRIPTION
**Pull request by domenicbove**
# Description

Breaks out 2 listeners (one internal for the brokers to talk on, one external for all other clients). You can set security mechanisms on both

Sasl scram now added as another sasl_protocol

Added multiple users for sasl plain

Scram and plain both use loops so additional users can easily be added

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Deployed with different combinations of security on each listener and then ran system test playbook

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: roles/confluent.variables/defaults/main.yml [hunk 1]
```diff
@@ -65,6 +56,49 @@ keystore_keypass: "{{ ssl_keystore_key_password if ssl_provided_keystore_and_tru
...
+  else 'none' }}
+kafka_broker_internal_listener_security_protocol: >-
+  {{ 'SASL_SSL' if kafka_broker_internal_listener_ssl_enabled|bool and kafka_broker_internal_listener_sasl_protocol_normalized in ['GSSAPI', 'PLAIN', 'SCRAM-SHA-256']
+  else 'SASL_PLAINTEXT' if not kafka_broker_internal_listener_ssl_enabled|bool and kafka_broker_internal_listener_sasl_protocol_normalized in ['GSSAPI', 'PLAIN', 'SCRAM-SHA-256']
+  else 'SSL' if kafka_broker_internal_listener_ssl_enabled|bool and kafka_broker_internal_listener_sasl_protocol == 'none'
+  else 'PLAINTEXT' }}
+
+# kafka_broker_internal_listener_security_protocol: >-
+#   {{ 'SASL_SSL' if kafka_broker_internal_listener_ssl_enabled|bool and kafka_broker_internal_listener_sasl_protocol_normalized == 'GSSAPI' or kafka_broker_internal_listener_sasl_protocol_normalized == 'PLAIN'
+#   else 'SASL_PLAINTEXT' if not kafka_broker_internal_listener_ssl_enabled|bool and kafka_broker_internal_listener_sasl_protocol_normalized == 'GSSAPI' or kafka_broker_internal_listener_sasl_protocol_normalized== 'PLAIN'
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml]: remove commented code

File: roles/confluent.variables/defaults/main.yml [hunk 2]
```diff
@@ -65,6 +56,49 @@ keystore_keypass: "{{ ssl_keystore_key_password if ssl_provided_keystore_and_tru
...
+  else 'SCRAM-SHA-256' if kafka_broker_external_listener_sasl_protocol|upper == 'SCRAM'
+  else 'PLAIN' if kafka_broker_external_listener_sasl_protocol|upper == 'PLAIN'
+  else 'none' }}
+kafka_broker_external_listener_security_protocol: >-
+  {{ 'SASL_SSL' if kafka_broker_external_listener_ssl_enabled|bool and kafka_broker_external_listener_sasl_protocol_normalized in ['GSSAPI', 'PLAIN', 'SCRAM-SHA-256']
+  else 'SASL_PLAINTEXT' if not kafka_broker_external_listener_ssl_enabled|bool and kafka_broker_external_listener_sasl_protocol_normalized in ['GSSAPI', 'PLAIN', 'SCRAM-SHA-256']
+  else 'SSL' if kafka_broker_external_listener_ssl_enabled|bool and kafka_broker_external_listener_sasl_protocol == 'none'
+  else 'PLAINTEXT' }}
+
+# TODO BUG here if sasl_protocol is kerberos -> gssapi
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml]: TODO still needs to be done?

File: roles/confluent.variables/defaults/main.yml [hunk 3]
```diff
@@ -89,3 +123,10 @@ control_center_ssl_enabled: "{{ssl_enabled}}"
 control_center_ssl_mutual_auth_enabled: "{{ ssl_mutual_auth_enabled }}"
 control_center_http_protocol: "{{ 'https' if control_center_ssl_enabled|bool else 'http' }}"
 control_center_port: 9021
+
+# sasl_scram_encryption_level: 512
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml]: more commented code

File: roles/confluent.kafka_broker/templates/server.properties.j2 [hunk 4]
```diff
@@ -9,43 +9,55 @@ broker.id={{ broker_id | default(groups.kafka_broker.index(inventory_hostname) +
...
 confluent.metrics.reporter.ssl.keystore.password={{keystore_storepass}}
 confluent.metrics.reporter.ssl.key.password={{keystore_keypass}}
 {% endif %}
-{% if sasl_protocol == 'plain' %}
+{% endif %}
+{% if kafka_broker_external_listener_sasl_protocol_normalized == 'PLAIN' %}
 confluent.metrics.reporter.sasl.mechanism=PLAIN
-confluent.metrics.reporter.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="client" password="client-secret";
+confluent.metrics.reporter.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required \
+   username="{{sasl_plain_users.admin.principal}}" password="{{sasl_plain_users.admin.password}}";
```
Comment (domenicbove) [inline roles/confluent.kafka_broker/templates/server.properties.j2]: @JumaX What user do you think the metrics reporter on the broker be calling in as?
I have it as admin rn for plain and scram

File: roles/confluent.variables/defaults/main.yml [hunk 5]
```diff
@@ -89,3 +116,54 @@ control_center_ssl_enabled: "{{ssl_enabled}}"
 control_center_ssl_mutual_auth_enabled: "{{ ssl_mutual_auth_enabled }}"
 control_center_http_protocol: "{{ 'https' if control_center_ssl_enabled|bool else 'http' }}"
 control_center_port: 9021
+
+# sasl_scram_admin_principal: admin
+# sasl_scram_admin_password: admin-secret
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml]: remove commented code

File: roles/confluent.kafka_broker/templates/server.properties.j2 (hunk 4 above)
Comment (JumaX) [reply roles/confluent.kafka_broker/templates/server.properties.j2]: @domenicbove good question.  Typically we see the customer use the same user as what they are using to authorize the Brokers.

**Pull request closed without merging**